engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java

```diff
@@ -14,13 +15,15 @@
 
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
+import org.enso.interpreter.dsl.Builtin;
 
 /** The main runtime type for Enso's Text. */
 @ExportLibrary(InteropLibrary.class)
 @ExportLibrary(TypesLibrary.class)
 public final class Text implements TruffleObject {
   private volatile Object contents;
   private volatile boolean isFlat;
+  private volatile long length = -1;
```

Comment:
Should `setContents` invalidate this?

Reply:
The `setContents` is only used to exchange one representation of the same content for another. E.g. the `length` has to stay the same. I put there `assert` to ensure that in https://github.com/enso-org/enso/pull/3864/commits/da66843a878d5bf02d4dde598362105feb1144f0